pkg/state/impl/etcd/etcd.go

```diff
@@ -0,0 +1,571 @@
+package etcd
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strconv"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/cosi-project/runtime/pkg/state/impl/store"
+	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+// State implements state.CoreState.
+type State struct {
+	cli       *clientv3.Client
+	marshaler store.Marshaler
+}
+
+// NewState creates new State with default options.
+func NewState(cli *clientv3.Client) *State {
+	return NewStateWithOptions(cli)
+}
+
+// NewStateWithOptions returns state builder function with options.
+func NewStateWithOptions(cli *clientv3.Client, opts ...StateOption) *State {
+	options := DefaultStateOptions()
+
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	return &State{
+		cli:       cli,
+		marshaler: store.ProtobufMarshaler{},
+	}
+}
+
+// Get a resource.
+func (st *State) Get(ctx context.Context, resourcePointer resource.Pointer, opts ...state.GetOption) (resource.Resource, error) { //nolint:ireturn
+	var options state.GetOptions
+
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	etcdKey := etcdKeyFromPointer(resourcePointer)
+
+	resp, err := st.cli.Get(ctx, etcdKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.Kvs) == 0 {
+		return nil, ErrNotFound(resourcePointer)
+	}
+
+	return st.unmarshalResource(resp.Kvs[0])
+}
+
+// List resources.
+func (st *State) List(ctx context.Context, resourceKind resource.Kind, opts ...state.ListOption) (resource.List, error) {
+	var options state.ListOptions
+
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	etcdKey := etcdKeyPrefixFromKind(resourceKind)
+
+	resp, err := st.cli.Get(ctx, etcdKey, clientv3.WithPrefix())
+	if err != nil {
+		return resource.List{}, err
+	}
+
+	resources := make([]resource.Resource, 0, len(resp.Kvs))
+
+	for _, kv := range resp.Kvs {
+		res, err := st.unmarshalResource(kv)
+		if err != nil {
+			return resource.List{}, err
+		}
+
+		if !options.LabelQuery.Matches(*res.Metadata().Labels()) {
+			continue
+		}
+
+		resources = append(resources, res)
+	}
+
+	return resource.List{
+		Items: resources,
+	}, nil
+}
+
+// Create a resource.
+func (st *State) Create(ctx context.Context, res resource.Resource, opts ...state.CreateOption) error {
+	resCopy := res.DeepCopy()
+
+	var options state.CreateOptions
+
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	etcdKey := etcdKeyFromPointer(resCopy.Metadata())
+
+	if err := resCopy.Metadata().SetOwner(options.Owner); err != nil {
+		return err
+	}
+
+	data, err := st.marshaler.MarshalResource(resCopy)
+	if err != nil {
+		return err
+	}
+
+	resp, err := st.cli.Txn(ctx).If(
+		clientv3.Compare(clientv3.Version(etcdKey), "=", 0), // not exists check
+	).Then(
+		clientv3.OpPut(etcdKey, string(data)),
+	).Commit()
+	if err != nil {
+		return err
+	}
+
+	if !resp.Succeeded {
+		return ErrAlreadyExists(resCopy.Metadata())
+	}
+
+	version, err := resource.ParseVersion("1")
```

Comment:
should we receive the version back in the response?

Reply:
done, changed the transaction a bit to do a get after put. did the same for update as well.